copy/copy.go

```diff
@@ -655,17 +725,122 @@ func (ic *imageCopier) copyLayer(ctx context.Context, srcInfo types.BlobInfo, po
+		// another method with a complex parameter list and semantics.
+		// It's sole purpose is to simplify the control with an early
+		// return if the image destination does not blob locking.
+		tryReusingBlob := func() (bool, types.BlobInfo, digest.Digest, error) {
+			reused, blobInfo, err := ic.c.dest.TryReusingBlob(ctx, srcInfo, ic.c.blobInfoCache, ic.canSubstituteBlobs)
+			if err != nil {
+				return false, types.BlobInfo{}, "", errors.Wrapf(err, "Error trying to reuse blob %s at destination", srcInfo.Digest)
+			}
+			if reused {
+				logrus.Debugf("Skipping blob %s (already present):", srcInfo.Digest)
+				bar.SetTotal(0, true)
+				ic.c.createProgressBar(pool, srcInfo, "blob", "skipped: already exists", mpb.BarReplaceOnComplete(bar))
+				return true, blobInfo, cachedDiffID, nil
+			}
+
+			if !ic.c.dest.SupportsBlobLocks() {
+				return false, types.BlobInfo{}, "", nil
+			}
+
+			// Check if another process might be copying the layer at the
+			// moment.  If so, wait for it's completion and try to reuse
+			// the blob again.
+			//
+			logrus.Debugf("Checking if blob %s is already being copied by another process", srcInfo.Digest)
+
+			// LockBlob() does not support try-lock semantics which would
+			// be a straight-forward mechanism to detect if another process
+			// is currently copying a specific layer.  However, with
+			// LockBlob() we are constrained by the file locks in
+			// containers/storage which cannot support try-lock semantics
+			// as fcntl(2) does not support that.
+			//
+			// The trick we are using here is to acquire the lock in a
+			// separate goroutine. The main routine first sleeps for
+			// the progressBarRefreshRate amount of time and will be
+			// signaled via a semaphore if and when the blob lock has
+			// been acquired.
+			blobLockSem := semaphore.NewWeighted(1)
+			blobLockSem.Acquire(ctx, 1)
+			var lockError error
+			go func() {
+				lockError = ic.c.dest.LockBlob(srcInfo)
+				blobLockSem.Release(1)
+			}()
+
+			copiedByAnotherProcess := false
+			if !blobLockSem.TryAcquire(1) {
+				// The initial sleep time is `progressBarRefreshRate`
+				// to make sure the bar is properly rendered before
+				// replacing it.
+				time.Sleep(progressBarRefreshRate)
+				if !blobLockSem.TryAcquire(1) {
+					logrus.Debugf("Blob %s is already being copied by another process", srcInfo.Digest)
+					copiedByAnotherProcess = true
+					// If we still don't have the lock, replace the
+					// initial progress bar with a new one
+					// indicating that the blob is copied by
+					// another process.
+					//
+					// Note that we call SetTotal() after the new
+					// bar is created to avoid potential flickering
+					// if the bar removal and replacement happens
+					// across refresh rates.
+					toReplaceBar := bar
+					bar = ic.c.createProgressBar(pool, srcInfo, "blob", "paused: being copied by another process", mpb.BarRemoveOnComplete(), mpb.BarReplaceOnComplete(bar))
```

Comment:
Is `bar` ever completed? If not, how does `pool.Wait` finish?

Reply:
It will complete. createProgressBar() will set the size to 0 if a filler is specified.

Reply:
If I am reading the code correctly, a zero-sized bar does not automatically complete; `toComplete` is set only by `Bar.SetTotal` or `Bar.IncrBy`.

Anyway, in the `!reused` case below, this progress bar should somehow be returned to the caller, so that it can be replaced before `copyLayerBlobFromStream`.

(In the `reused` case, this progress bar is replaced, which completes it.)